modules/core/src/main/scala/iota/internal/toolbelts.scala

```diff
@@ -109,81 +57,91 @@ object IotaReflectiveToolbelt {
     apply(runtimeUniverse)
 }
 
-class IotaReflectiveToolbelt[U <: Universe](override val u: U)
-    extends IotaCommonToolbelt { override type Uu = U }
+final class IotaReflectiveToolbelt[U <: Universe](val u: U)
+    extends Toolbelt
+    with TypeListTrees
+    with TypeListParsers
+    with TypeListEvaluators
+    with TypeListBuilders
+    with CoproductAPI { override type Uu = U }
 
-private[internal] sealed abstract class IotaCommonToolbelt {
 
-  type Uu <: Universe
-  val u: Uu
+private[internal] class IotaMacroToolbelt[C <: Context](val c: C)
+    extends MacroToolbelt
+    with TypeListTrees
+    with TypeListParsers
+    with TypeListEvaluators
+    with TypeListBuilders
+    with CoproductAPI
+    with MacroAPI { override type Cc = C }
 
-  import u._
+private[internal] object IotaMacroToolbelt {
+  def apply[C <: Context](c: C): IotaMacroToolbelt[c.type] =
+    new IotaMacroToolbelt[c.type](c)
 
-  object tree {
-    sealed trait NodeF[A]
-    case class Cons[A](head: Type, tail: A) extends NodeF[A]
-    case class Concat[A](nodes: List[A]) extends NodeF[A]
-    case class Reverse[A](node: A) extends NodeF[A]
-    case class Take[A](n: Int, node: A) extends NodeF[A]
-    case class Drop[A](n: Int, node: A) extends NodeF[A]
-    case class NNil[A]() extends NodeF[A]
-
-    object Concat { def apply[A](nodes: A*): Concat[A] = Concat(nodes.toList) }
-    object Cons {
-      def apply[T]: TPartiallyApplied[T] = new TPartiallyApplied[T]
-
-      class TPartiallyApplied[T] {
-        def apply[A](tail: A)(implicit evT: WeakTypeTag[T]): Cons[A] = Cons[A](evT.tpe, tail)
-      }
+  final class Cache {
```

Comment:
It's used by the actual macro methods for memoizing the calculated `TList/KList` types.